### PR TITLE
Fix RefNull issues

### PR DESCRIPTION
**Pull request by dcodeIO**
Fixes https://github.com/WebAssembly/binaryen/issues/3119

### CODE REVIEW

File: src/ir/ExpressionAnalyzer.cpp [hunk 1]
```diff
@@ -219,7 +219,7 @@ template<typename T> void visitImmediates(Expression* curr, T& visitor) {
       visitor.visitInt(curr->op);
       visitor.visitNonScopeName(curr->nameOperand);
     }
-    void visitRefNull(RefNull* curr) {}
+    void visitRefNull(RefNull* curr) { visitor.visitType(curr->type); }
```
Comment (dcodeIO) [inline src/ir/ExpressionAnalyzer.cpp:222]: This was the culprit, leading to `(ref.null any)` and `(ref.null func)` for example to compare equal, leading to invalid code folding.

File: src/passes/Precompute.cpp [hunk 2]
```diff
@@ -134,9 +134,12 @@ struct Precompute
...
           }
-        } else if (singleValue.isNull() &&
-                   curr->value->template is<RefNull>()) {
-          return;
+        } else if (singleValue.isNull()) {
+          if (auto* n = curr->value->template dynCast<RefNull>()) {
+            n->finalize(singleValue.type);
+            curr->finalize();
+            return;
+          }
```
Comment (dcodeIO) [inline src/passes/Precompute.cpp:142]: While on it I also looked at all the uses of `RefNull` again and found that this one might be problematic as well?

File: src/tools/execution-results.h [hunk 3]
```diff
@@ -73,7 +73,7 @@ struct ExecutionResults {
           // change (after duplicate function elimination or roundtripping)
           // while the function contents are still the same
           for (Literal& val : ret) {
-            if (val.type == Type::funcref) {
+            if (val.type == Type::funcref && !val.isNull()) {
```
Comment (dcodeIO) [inline src/tools/execution-results.h:76]: Another problem I found during debugging, where `(ref.null func)` was replaced with `(ref.func funcref)`, not accounting for `null`.

File: src/tools/execution-results.h [hunk 4]
```diff
@@ -112,7 +112,8 @@ struct ExecutionResults {
       }
       std::cout << "[fuzz-exec] comparing " << name << '\n';
       if (results[name] != other.results[name]) {
-        std::cout << "not identical!\n";
+        std::cout << "not identical! " << results[name]
+                  << " != " << other.results[name] << "\n";
```
Comment (dcodeIO) [inline src/tools/execution-results.h:116]: Unrelated, but helped me during debugging so I left it in.

**Comment by dcodeIO** [comment]
```
Invocations so far:
   FuzzExec: 27560
   CompareVMs: 7583
   CheckDeterminism: 2391
   Wasm2JS: 6672
   Asyncify: 7045

ITERATION: 32357
```

File: test/passes/optimize-instructions_all-features.txt [hunk 5]
```diff
@@ -3713,7 +3713,11 @@
   )
  )
  (func $if-arms-subtype (result anyref)
-  (ref.null extern)
+  (if (result anyref)
+   (i32.const 0)
+   (ref.null extern)
+   (ref.null func)
```
Comment (aheejin) [inline test/passes/optimize-instructions_all-features.txt:3719]: Currently this is identical to the original function in optimize-instruction_all-features.wast, so it does not serve any testing purpose.

I suggest to change `if-arms-subtype` function in optimize-instructions_all-feature.wast to this:
```wat
(func $if-arms-subtype (result anyref)
  (if (result anyref)
    (i32.const 0)
    (ref.null extern)
    (ref.null extern)
  )
)
```

---

Background:
This test was first added in #2451 so that when `if`'s true body and false body are identical, it is possible to fold the expression to if's true body even if the `if`'s true body's type is a subtype of `if`'s type. The relevant optimizing routine is [here](https://github.com/WebAssembly/binaryen/blob/0c53fb5d84afe32bfdf13231908aad760cb1d89d/src/passes/OptimizeInstructions.cpp#L559-L590).

So, before #2451, only this was able to be folded:
```wat
(if (result i32)
  (i32.const 0)
  (i32.const 3)
  (i32.const 3)
)
```
Note that `if`'s true body (and also false body)'s type `i32` is the same as the `if`'s type.

#2451 allowed this to be folded too:
```wat
(if (result anyref)
  (i32.const 0)
  (ref.null) ;; nullref type
  (ref.null) ;; nullref type
)
```
Then `ref.null` had `nullref` type, which used to be a subtype of `anyref`. So now we were able to fold even if `if`'s bodies are subtype (`nullref`) of `if`'s type (`anyref`).

Note that this optimization only happens when `if`'s true body and false body are [identical](https://github.com/WebAssembly/binaryen/blob/0c53fb5d84afe32bfdf13231908aad760cb1d89d/src/passes/OptimizeInstructions.cpp#L560). But it looks like #3084 mistakenly changed this code and made these to bodies not identical:
```wat
(if (result anyref)
  (i32.const 0)
  (ref.null extern)
  (ref.null func)
)
```
So now these bodies are not identical, this should not have been optimized. But this was still optimized and went unnoticed because of the bug you fixed in this PR.

So, long story short, we should make them identical again. :)

**Review by aheejin — APPROVED** [review]

File: test/passes/optimize-instructions_all-features.txt (hunk 5 above)
Comment (aheejin) [reply test/passes/optimize-instructions_all-features.txt:3719]: And to make it more readable, it might be better to add a short comment for `$if-arms-subtype` in optimize-instruction_all-features.wast, something like
```
This function tests if 'if' can be folded as its true body when
1. It's true and false body are identical
2. When if's true (and false) body's type is not identical but a subtype of 'if''s type
```

Comment (tlively) [reply test/passes/optimize-instructions_all-features.txt:3719]: Maybe in addition to the case you suggested, we can also keep the current one that is the same in the wast and the output as a negative test showing that the arms are correctly not merged.

**Review by tlively — APPROVED** [review]
LGTM with @aheejin's comment about the test addressed.
